src/sources/docker.rs

```diff
@@ -263,97 +283,170 @@ impl DockerSource {
 
-                    let id = ContainerId::new(container.id);
+                let id = ContainerId::new(id);
+                self.containers.insert(id.clone(), self.esb.start(id));
+            });
 
-                    self.containers.insert(id.clone(), self.esb.start(id));
+        Ok(self)
+    }
+
+    async fn run(mut self) {
+        loop {
+            tokio::select! {
+                value = self.main_recv.next() => {
+                    match value {
+                        Some(info) => {
+                            let state = self
+                                .containers
+                                .get_mut(&info.id)
+                                .expect("Every ContainerLogInfo has it's ContainerState");
+                            if state.return_info(info) {
+                                self.esb.restart(state);
+                            }
+                        }
+                        None => {
+                            error!(message = "docker source main stream has ended unexpectedly");
+                            info!(message = "Shuting down docker source");
+                            return;
+                        }
+                    };
                 }
+                value = self.events.next() => {
+                    match value {
+                        Some(Ok(mut event)) => {
+                            let action = event.action.unwrap();
+                            let actor = event.actor.take().unwrap();
+                            let id = actor.id.unwrap();
+                            let attributes = actor.attributes.unwrap();
+
+                            trace!(
+                                message = "docker event",
+                                id = field::display(&id),
+                                action = field::display(&action),
+                                timestamp = field::display(event.time.unwrap()),
+                                attributes = field::debug(&attributes),
+                            );
+                            let id = ContainerId::new(id);
+
+                            // Update container status
+                            match action.as_str() {
+                                "die" | "pause" => {
+                                    if let Some(state) = self.containers.get_mut(&id) {
+                                        state.stoped();
+                                    }
+                                }
+                                "start" | "upause" => {
+                                    if let Some(state) = self.containers.get_mut(&id) {
+                                        state.running();
+                                        self.esb.restart(state);
+                                    } else {
+                                        // This check is necessary since docker api doesn't have way to include
+                                        // names into request.
```

Comment:
you should be able to use a `filter` hashmap to narrow down calls to the `events` API